torch/csrc/autograd/engine.cpp

```diff
@@ -651,15 +674,25 @@ auto Engine::execute(const edge_list& roots,
       return graph_task.outstanding_tasks_.load() == 0;
     });
   } else {
-    // Get back to work while we wait for our new graph_task to
-    // complete!
-    // See Note [Reentrant backwards]
-    // If no extra threads remaining, create a new thread for reentrant call
     graph_task.owner_ = worker_device;
-    add_thread_pool_task(&graph_task);
-    graph_task.not_done_.wait(lock, [&graph_task]{
-      return graph_task.outstanding_tasks_.load() == 0;
-    });
+    ++total_depth;
+    graph_task.reentrant_depth_ = total_depth;
+    if(current_depth >= max_recursion_depth_){
+      // See Note [Reentrant backwards]
+      // If reached the max depth, switch to a different thread
+      add_thread_pool_task(&graph_task);
```

Comment:
What will happen if we exceed thread_pool_shared_->graphtasks_queue_.size() here?

Reply:
std::queue should resize dynamically, or am I misunderstanding your question?